contributor-docs/adrs/adr-004-children-as-api.md

```diff
@@ -0,0 +1,403 @@
+
+   ```jsx
+   <Flash variant="success">
+     <CheckIcon /> Changes saved!
+   </Flash>
+   <Flash variant="danger">
+     <AlertIcon /> Your changes were not saved!
+   </Flash>
+   ```
+
+2. You can bring your own icon components, the component does not depend on a specific version of octicons.
+3. When a product team wants to explore a use cases which isn’t baked into the component, they are not blocked by our team.
+
+   Example:
+
+   <img width="375" alt="flash with icon and close" src="https://user-images.githubusercontent.com/1863771/144945209-1480b9f5-895d-458d-8c68-4ce7b6d6b7b4.png">
+
+   ```jsx
+   <Flash variant="success" sx={{display: 'flex', justifyContent: 'space-between'}}>
+     <span>
+       <CheckIcon /> Changes saved!
+     </span>
+     <Button variant="invisible" icon={CheckIcon} aria-label="Hide flash message" onClick={onDismiss} />
+   </Flash>
+   ```
+
+<br/>
+
+### Tradeoffs of this approach / When not to use children
+
+Our goal isn't to put all content inside children though. Composition offers flexibility to the consumer of the component, this flexibility, however, can also lead to inconsistencies.
+
+<img width="375" alt="image 6" src="https://user-images.githubusercontent.com/1863771/144945216-6237084d-3d97-45c7-9505-6aed473afb03.png">
+
+```jsx
+<Flash variant="success">
+  // oh oh, we don't know if that color or icon size is the right choice!
+  <CheckIcon size={20} color="success.emphasis" /> Changes saved!
+</Flash>
+
+<Flash variant="success">
+  <CheckIcon /> Changes saved!
+</Flash>
+```
+
+_Note: We need to assume good intent here, developers using the components aren’t trying to break the system. They are either trying to implement something that the system’s happy path does not support OR there are multiple ways of doing something with primer and they have unintentionally picked the approach that is not recommended._
+
+<br/>
+
+The general wisdom is to _Make the right (recommended) thing easy to do and the wrong (not recommended) hard to do_. When going off the happy path, developers should feel some friction, some weight, code that “feels hacky” or feels like a workaround.
+
+In the above case, if we want to make the recommended path easier and other paths harder, we can change the API to look something like this -
+
+```jsx
+<Flash variant="success" icon={CheckIcon}>
+  Changes saved!
+</Flash>
+```
+
+We are still using `children` for text content, but we have moved the `icon` back as a prop with reduced flexibility.
+
+When intentionally going off the happy path, developers can still drop down an abstraction level add an `Icon` to `children`, they would have to pick up the additional effort of setting compatible color, size and margin themselves. However, when it’s unintentional, it would feel like way too much work that the component should be doing automatically.
```

Comment:
```suggestion
When intentionally going off the happy path, developers can still drop down an abstraction level to add an `Icon` to `children`, though they would have to pick up the additional effort of setting compatible color, size and margin themselves. However, when it’s unintentional, it would feel like way too much work that the component should be doing automatically.
```

Feel free to edit my suggestion as needed if I didn't hit the right tone, just wanted to try to clear up some confusion from the wording.